In [1]:
import sklearn
print(sklearn.__version__)

1.0


In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# 손글씨 데이터셋 불러오기
digit = load_digits()

In [6]:
# 각각의 데이터의 feature 값 저장
digit_data = digit.data
# 데이터 라벨링
digit_label = digit.target
# 라벨의 이름 표시
digit.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [7]:
# 자세한 데이터셋 설명
print(digit.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

__사진의 개수: 1797, 사진 하나의 데이터 길이: 64, Attribute Information: 8x8 image of integer pixels in the range 0..16.__

In [90]:
# 학습시킬 데이터와 test할 데이터 나누기
X_train, X_test, y_train, y_test = train_test_split(digit_data, digit_label, test_size = 0.2, random_state = 15)

# 데이터의 분류 작업 기법과 학습시킨 후 평가 방법 import
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix


# accuracy_score 비교
decision_tree = DecisionTreeClassifier(random_state = 15)
decision_tree.fit(X_train, y_train)
y_pred_decision = decision_tree.predict(X_test)

In [87]:
print(accuracy_score(y_test, y_pred_decision))

0.9444444444444444


In [86]:
random_forest = RandomForestClassifier(random_state=15)
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(accuracy_score(y_test, y_pred_random_forest))

0.9444444444444444


In [98]:
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)

print(accuracy_score(y_test, y_pred_svm))

0.868421052631579


In [85]:
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred_sgd = sgd_model.predict(X_test)

print(accuracy_score(y_test, y_pred_sgd))

0.6111111111111112


In [84]:
logistic_model = LogisticRegression(max_iter=10**5)
logistic_model.fit(X_train, y_train)
y_pred_logistic = logistic_model.predict(X_test)

print(accuracy_score(y_test, y_pred_logistic))

0.9444444444444444


In [88]:
# 와인 데이터셋 불러오기
from sklearn.datasets import load_wine

wine = load_wine()
wine_data = wine.data
wine_label = wine.target

print(wine.target_names)

['class_0' 'class_1' 'class_2']


In [89]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

__Number of Instances: 178 (50 in each of three classes), Number of Attributes: 13 numeric, predictive attributes and the class
- __Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline
- __class:
            - class_0
            - class_1
            - class_2
            
__Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            10.6  30.0    19.5   3.3
    Magnesium:                    70.0 162.0    99.7  14.3
    Total Phenols:                0.98  3.88    2.29  0.63
    Flavanoids:                   0.34  5.08    2.03  1.00
    Nonflavanoid Phenols:         0.13  0.66    0.36  0.12
    Proanthocyanins:              0.41  3.58    1.59  0.57
    Colour Intensity:              1.3  13.0     5.1   2.3
    Hue:                          0.48  1.71    0.96  0.23
    OD280/OD315 of diluted wines: 1.27  4.00    2.61  0.71
    Proline:                       278  1680     746   315
    ============================= ==== ===== ======= =====

In [82]:
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state = 10)

decision_tree = DecisionTreeClassifier(random_state = 10)
decision_tree.fit(X_train, y_train)
y_pred_decision = decision_tree.predict(X_test)

print(accuracy_score(y_test, y_pred_decision))

0.9444444444444444


In [81]:
random_forest = RandomForestClassifier(random_state=15)
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(accuracy_score(y_test, y_pred_random_forest))

0.9298245614035088


In [80]:
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)

print(accuracy_score(y_test, y_pred_svm))

0.868421052631579


In [79]:
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred_sgd = sgd_model.predict(X_test)

print(accuracy_score(y_test, y_pred_sgd))

0.8245614035087719


In [78]:
logistic_model = LogisticRegression(max_iter = 10**5)
logistic_model.fit(X_train, y_train)
y_pred_logistic = logistic_model.predict(X_test)

print(accuracy_score(y_test, y_pred_logistic))

0.8947368421052632


In [92]:
# breast_cancer 데이터셋 불러오기
from sklearn.datasets import load_breast_cancer
from sklearn.metrics import recall_score

breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data
breast_cancer_label = breast_cancer.target

print(breast_cancer.target_names)

['malignant' 'benign']


In [59]:
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

__Number of Instances: 569, Number of Attributes: 30 numeric, predictive attributes and the class

__Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)
        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radius, field
        10 is Radius SE, field 20 is Worst Radius.
        - class:
                - WDBC-Malignant
                - WDBC-Benign
- __Summary Statistics:

    ===================================== ====== ======
                                           Min    Max
    ===================================== ====== ======
    radius (mean):                        6.981  28.11
    texture (mean):                       9.71   39.28
    perimeter (mean):                     43.79  188.5
    area (mean):                          143.5  2501.0
    smoothness (mean):                    0.053  0.163
    compactness (mean):                   0.019  0.345
    concavity (mean):                     0.0    0.427
    concave points (mean):                0.0    0.201
    symmetry (mean):                      0.106  0.304
    fractal dimension (mean):             0.05   0.097
    radius (standard error):              0.112  2.873
    texture (standard error):             0.36   4.885
    perimeter (standard error):           0.757  21.98
    area (standard error):                6.802  542.2
    smoothness (standard error):          0.002  0.031
    compactness (standard error):         0.002  0.135
    concavity (standard error):           0.0    0.396
    concave points (standard error):      0.0    0.053
    symmetry (standard error):            0.008  0.079
    fractal dimension (standard error):   0.001  0.03
    radius (worst):                       7.93   36.04
    texture (worst):                      12.02  49.54
    perimeter (worst):                    50.41  251.2
    area (worst):                         185.2  4254.0
    smoothness (worst):                   0.071  0.223
    compactness (worst):                  0.027  1.058
    concavity (worst):                    0.0    1.252
    concave points (worst):               0.0    0.291
    symmetry (worst):                     0.156  0.664
    fractal dimension (worst):            0.055  0.208
    ===================================== ====== ======__

In [93]:
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, breast_cancer_label, test_size=0.2, random_state=15)

#confusion matrix and recall score 비교

decision_tree = DecisionTreeClassifier(random_state = 15)
decision_tree.fit(X_train, y_train)
y_pred_decision = decision_tree.predict(X_test)


print(confusion_matrix(y_test, y_pred_decision))
print(recall_score(y_test, y_pred_decision))

[[34  5]
 [ 1 74]]
0.9866666666666667


In [94]:
random_forest = RandomForestClassifier(random_state=15)
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(confusion_matrix(y_test, y_pred_random_forest))
print(recall_score(y_test, y_pred_random_forest))

[[33  6]
 [ 2 73]]
0.9733333333333334


In [95]:
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)

print(confusion_matrix(y_test, y_pred_svm))
print(recall_score(y_test, y_pred_svm))

[[26 13]
 [ 2 73]]
0.9733333333333334


In [96]:
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred_sgd = sgd_model.predict(X_test)

print(confusion_matrix(y_test, y_pred_sgd))
print(recall_score(y_test, y_pred_sgd))

[[32  7]
 [ 6 69]]
0.92


In [ ]:
logistic_model = LogisticRegression(max_iter = 10**5)
logistic_model.fit(X_train, y_train)
y_pred_logistic = logistic_model.predict(X_test)

print(confusion_matrix(y_test, y_pred_logistic))
print(recall_score(y_test, y_pred_logistic))

# 회고
## 1. 이번 프로젝트에서 어려웠던 점
__Precision__ 과 __Recall__ 에 대해 개념이 헷갈렸었다.

## 2. 프로젝트를 진행하면서 알아낸 점 혹은 아직 모호한 점
Precision은 __positive__ 라고 한 데이터 중에 실제 __positive__ 의 비율이고, Recall은 실제 __positive__ 데이터 중에서 __positive__ 라고 한 비율이다.

## 3. 루브릭 평가 지표를 맞추기 위해 시도한 것들
- 3가지 데이터셋에 대해 모델의 평가지표가 적절히 선택되었는가?
- __*cancer의 음성, 양성 판단은 양성이 음성으로 나오면 안되므로 'Recall'을 통해 성능을 검증해야 하므로 'confusion_matrix'를 사용하여 'recall_score'를 구한다. digits와 wine의 경우는 정확도가 중요하기 때문에 'accuracy'를 사용한다.*__